In [1]:
import sys
print(sys.executable)

c:\Users\kirill\AppData\Local\anaconda3\envs\iei-integration\python.exe


In [2]:
import pandas as pd
print(pd.__version__)

3.0.6


In [3]:
from pathlib import Path
Path.cwd()

WindowsPath('c:/Users/kirill/Desktop/IEI Integrations/fuentes de data/notebooks')

In [5]:
import pandas as pd
## data\datos-abiertos-csv-segundo-trimestre26.json
df = pd.read_json("../data/datos-abiertos-csv-segundo-trimestre26.json")
df.head()

,nif,denominacion,numero_registro,direccion_tipo_via,direccion_nombre_via,direccion_numero,direccion_complemento,direccion_provincia_id,direccion_provincia_nombre,direccion_municipio_id,direccion_municipio_nombre,direccion_isla_id,direccion_isla_nombre,telefono_1,telefono_2,pagina_web,area_id,area_nombre,subarea_id,subarea_nombre
0,G35068824,Asociacion Adepsi,ERE1986CA00001,Calle,Lomo La Plana,28,_U,35,Las Palmas,35001,Palmas de Gran Canaria (Las),GC,Gran Canaria,(34)928414484,_U,wwwadepsi.org,3,Discapacidad,_U,_U
1,G35103431,Radio Ecca Fundacion Canaria,ERE1986CA00002,Avda,Avenida De Escaleritas 64,64,1,35,Las Palmas,35001,Palmas de Gran Canaria (Las),GC,Gran Canaria,(34)928289168,_U,_U,7,Centros Ocupacionales,_U,_U
2,R3500084C,Ciudad De San Juan De Dios,ERE1987CA00001,Ctra,Carretera Ciudad San Juan De Dios,1,_U,35,Las Palmas,35001,Palmas de Gran Canaria (Las),GC,Gran Canaria,(34)928339080,_U,www.sjdlaspalmas.es,3,Discapacidad,_U,_U
3,G35025048,Civitas Centros Especializados En Personas Con...,ERE1987CA00002,Calle,La Calzada,88,_U,35,Las Palmas,35001,Palmas de Gran Canaria (Las),GC,Gran Canaria,(34)928350138,_U,www.asociacioncivitas.es,3,Discapacidad,_U,_U
4,G35053248,Asociación De Pensionistas Jubilados Por Vejez...,ERE1987CA00003,Calle,Los Martínez De Escobar,16,Izq A,35,Las Palmas,35001,Palmas de Gran Canaria (Las),GC,Gran Canaria,(34)928263023,_U,_U,2,Tercera Edad,_U,_U


In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 810 entries, 0 to 809
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype
---  ------                      --------------  -----
 0   nif                         810 non-null    str  
 1   denominacion                810 non-null    str  
 2   numero_registro             810 non-null    str  
 3   direccion_tipo_via          810 non-null    str  
 4   direccion_nombre_via        810 non-null    str  
 5   direccion_numero            810 non-null    str  
 6   direccion_complemento       810 non-null    str  
 7   direccion_provincia_id      810 non-null    str  
 8   direccion_provincia_nombre  810 non-null    str  
 9   direccion_municipio_id      810 non-null    int64
 10  direccion_municipio_nombre  810 non-null    str  
 11  direccion_isla_id           810 non-null    str  
 12  direccion_isla_nombre       810 non-null    str  
 13  telefono_1                  810 non-null    str  
 14  telefono_2           

In [7]:
df.nunique()

nif                           745
denominacion                  745
numero_registro               749
direccion_tipo_via             23
direccion_nombre_via          685
direccion_numero              133
direccion_complemento         180
direccion_provincia_id         13
direccion_provincia_nombre     13
direccion_municipio_id         90
direccion_municipio_nombre     84
direccion_isla_id               8
direccion_isla_nombre           8
telefono_1                    728
telefono_2                    171
pagina_web                    173
area_id                         9
area_nombre                     9
subarea_id                     17
subarea_nombre                 17
dtype: int64

In [8]:
df.isna().sum()

nif                           0
denominacion                  0
numero_registro               0
direccion_tipo_via            0
direccion_nombre_via          0
direccion_numero              0
direccion_complemento         0
direccion_provincia_id        0
direccion_provincia_nombre    0
direccion_municipio_id        0
direccion_municipio_nombre    0
direccion_isla_id             0
direccion_isla_nombre         0
telefono_1                    0
telefono_2                    0
pagina_web                    0
area_id                       0
area_nombre                   0
subarea_id                    0
subarea_nombre                0
dtype: int64

In [9]:
col = "nif"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))


dtype: str
missing: 0
non-null: 810
unique: 745


0    G35068824
1    G35103431
2    R3500084C
3    G35025048
4    G35053248
5    G35347814
6    G35236918
7    G35065911
8    G38030581
9    G28821254
Name: nif, dtype: str

nif
G76322395    7
G72405277    5
G38728382    3
G76262872    3
G02794337    3
G35141365    3
G76365410    3
G78919271    3
G35963875    2
G35229996    2
G76802248    2
G76225812    2
G86963659    2
G76256346    2
G38778338    2
G38296257    2
G06835649    2
G02855898    2
G38884789    2
G35763945    2
Name: count, dtype: int64

In [10]:
s = df["nif"]

print("Lengths:")
display(s.str.len().value_counts())

print("Placeholder _U:", s.eq("_U").sum())
print("Blank values:", s.str.strip().eq("").sum())
print("Leading/trailing whitespace:", s.ne(s.str.strip()).sum())

counts = s.value_counts()
print("Distinct NIFs appearing more than once:", counts.gt(1).sum())

display(
    df.loc[
        s.duplicated(keep=False),
        ["nif", "denominacion", "numero_registro",
         "area_nombre", "subarea_nombre"]
    ].sort_values("nif")
)

Lengths:


nif
9    810
Name: count, dtype: int64

Placeholder _U: 0
Blank values: 0
Leading/trailing whitespace: 0
Distinct NIFs appearing more than once: 51


,nif,denominacion,numero_registro,area_nombre,subarea_nombre
676,G01721471,Asoc Comunidad Peruana En Tenerife Copeten,ERE2023TF00129,Varios,Asociaciones Culturales Deportivas SociO-Educa...
677,G01721471,Asoc Comunidad Peruana En Tenerife Copeten,ERE2023TF00129,Varios,Asociación De Igualdad Lgtbi
600,G02794337,Fundacion Maroparque Fundacion Canaria,ERE2022TF00114,Voluntariado,Protección Del Medio Ambiente Y Defensa Del Me...
599,G02794337,Fundacion Maroparque Fundacion Canaria,ERE2022TF00114,Voluntariado,Educación Ciencia Cultura Deportes Y Patrimoni...
598,G02794337,Fundacion Maroparque Fundacion Canaria,ERE2022TF00114,Voluntariado,Cualquier Otro Campo De Actuación En El Que Se...
...,...,...,...,...,...
471,G86963659,Lares Asociacion Union De Residencias Y Servic...,ERE2021TF00119,Voluntariado,Cualquier Otro Campo De Actuación En El Que Se...
702,G90196072,Fundacion Bosco Global,ERE2024GC00026,Varios,Comedores Sociales Casas Hogar Comunidades Rel...
701,G90196072,Fundacion Bosco Global,ERE2024GC00026,Varios,Otros
793,R2802790B,Fundacion Hospitalarias,ERE2026CA00005,Varios,Comedores Sociales Casas Hogar Comunidades Rel...


In [11]:
df.groupby("nif")["denominacion"].nunique().value_counts()

denominacion
1    745
Name: count, dtype: int64

In [12]:
df.groupby("nif")["numero_registro"].nunique().value_counts()

numero_registro
1    741
2      4
Name: count, dtype: int64

In [13]:
df.groupby("numero_registro")["denominacion"].nunique().value_counts()

denominacion
1    749
Name: count, dtype: int64

In [14]:
df.groupby("numero_registro")["nif"].nunique().value_counts()

nif
1    749
Name: count, dtype: int64

In [15]:
df.groupby("denominacion")["nif"].nunique().value_counts()

nif
1    745
Name: count, dtype: int64

In [16]:
df.groupby("denominacion")["numero_registro"].nunique().value_counts()

numero_registro
1    741
2      4
Name: count, dtype: int64

In [ ]:
registrations = df.groupby("nif")["numero_registro"].unique()

lengths = registrations.str.len()
mask = lengths > 1

display(registrations[mask])


nif
G35141365    [ERE2022TF00115, ERE2024TF00037]
G35963875    [ERE2021TF00012, ERE2021TF00153]
G38296257    [ERE2021TF00131, ERE2021TF00143]
G76365410    [ERE2023GC00003, ERE2024GC00007]
Name: numero_registro, dtype: object

In [22]:
registrations_per_nif = df.groupby("nif")["numero_registro"].nunique()
multiple_registration_nifs = registrations_per_nif[
    registrations_per_nif > 1
].index

with pd.option_context("display.max_colwidth", None):
    display(
        df.loc[
            df["nif"].isin(multiple_registration_nifs),
            ["nif", "denominacion", "numero_registro"]
        ]
        ##.drop_duplicates()
        .sort_values(["nif", "numero_registro"])
    )

,nif,denominacion,numero_registro
601,G35141365,Fundacion Canaria Centro De Orientacion Familiar De Canaria,ERE2022TF00115
602,G35141365,Fundacion Canaria Centro De Orientacion Familiar De Canaria,ERE2022TF00115
711,G35141365,Fundacion Canaria Centro De Orientacion Familiar De Canaria,ERE2024TF00037
417,G35963875,Fundación Canaria Global Iniciativas Sociales,ERE2021TF00012
506,G35963875,Fundación Canaria Global Iniciativas Sociales,ERE2021TF00153
487,G38296257,Asoc. Feminista Mercedes Machado,ERE2021TF00131
498,G38296257,Asoc. Feminista Mercedes Machado,ERE2021TF00143
621,G76365410,Asoc Las Palmas Invita,ERE2023GC00003
687,G76365410,Asoc Las Palmas Invita,ERE2024GC00007
688,G76365410,Asoc Las Palmas Invita,ERE2024GC00007


In [26]:
nifs = df.loc[
            df["nif"].isin(multiple_registration_nifs),
            ["nif", "denominacion", "numero_registro"]
        ].sort_values(["nif", "numero_registro"])["nif"]
nifs

601    G35141365
602    G35141365
711    G35141365
417    G35963875
506    G35963875
487    G38296257
498    G38296257
621    G76365410
687    G76365410
688    G76365410
Name: nif, dtype: str

In [44]:
def highlight_differences(column):
    reference = column.iloc[0]
    return [
        "background-color: #440559" if value != reference else ""
        for value in column
    ]

for col in nifs.drop_duplicates():
    selected = df.loc[df["nif"] == col]
    distinct_counts = selected.nunique(dropna=False)
    different_columns = distinct_counts[distinct_counts > 1].index
    print("NIF: ", col)
    display(selected[different_columns].style.apply(highlight_differences, axis=0))

NIF:  G35141365


,numero_registro,direccion_nombre_via,direccion_numero,area_id,area_nombre,subarea_id,subarea_nombre
601,ERE2022TF00115,Calle Luis Saavedra Miranda,8,6,Exclusión Social,_U,_U
602,ERE2022TF00115,Calle Luis Saavedra Miranda,8,9,Voluntariado,2,Servicios Sociales
711,ERE2024TF00037,Don Pio Coronado,118,3,Discapacidad,_U,_U


NIF:  G35963875


,numero_registro,direccion_complemento,pagina_web
417,ERE2021TF00012,1,_U
506,ERE2021TF00153,5º,www.globaliniciativassociales.org


NIF:  G38296257


,numero_registro,direccion_nombre_via,direccion_numero,direccion_complemento
487,ERE2021TF00131,De La Libertad Vviendas Locales 7 Y 8.,18,_U
498,ERE2021TF00143,La Libertad,15,7 Y 8


NIF:  G76365410


,numero_registro,direccion_tipo_via,direccion_nombre_via,pagina_web,area_id,area_nombre,subarea_id,subarea_nombre
621,ERE2023GC00003,Calle,Pinar De Inagua 38 - Vivienda 25,laspalmasinvita.org,8,Varios,I,Otros
687,ERE2024GC00007,,Calle Pinar De Inagua,_U,8,Varios,B,Comedores Sociales Casas Hogar Comunidades Religiosas Talleres Sociales Desarrollo Social E Inserción Social Y Fomento Y Empleo
688,ERE2024GC00007,,Calle Pinar De Inagua,_U,2,Tercera Edad,_U,_U


In [27]:
selected = df.loc[df["nif"] == "G35141365"]
selected

,nif,denominacion,numero_registro,direccion_tipo_via,direccion_nombre_via,direccion_numero,direccion_complemento,direccion_provincia_id,direccion_provincia_nombre,direccion_municipio_id,direccion_municipio_nombre,direccion_isla_id,direccion_isla_nombre,telefono_1,telefono_2,pagina_web,area_id,area_nombre,subarea_id,subarea_nombre
601,G35141365,Fundacion Canaria Centro De Orientacion Famili...,ERE2022TF00115,Calle,Calle Luis Saavedra Miranda,8,_U,35,Las Palmas,35001,Palmas de Gran Canaria (Las),GC,Gran Canaria,(34)928200006,(34)928200006,www.fundacioncof.org,6,Exclusión Social,_U,_U
602,G35141365,Fundacion Canaria Centro De Orientacion Famili...,ERE2022TF00115,Calle,Calle Luis Saavedra Miranda,8,_U,35,Las Palmas,35001,Palmas de Gran Canaria (Las),GC,Gran Canaria,(34)928200006,(34)928200006,www.fundacioncof.org,9,Voluntariado,2,Servicios Sociales
711,G35141365,Fundacion Canaria Centro De Orientacion Famili...,ERE2024TF00037,Calle,Don Pio Coronado,118,_U,35,Las Palmas,35001,Palmas de Gran Canaria (Las),GC,Gran Canaria,(34)928200006,(34)928200006,www.fundacioncof.org,3,Discapacidad,_U,_U


In [28]:
distinct_counts = selected.nunique(dropna=False)
display(distinct_counts)

nif                           1
denominacion                  1
numero_registro               2
direccion_tipo_via            1
direccion_nombre_via          2
direccion_numero              2
direccion_complemento         1
direccion_provincia_id        1
direccion_provincia_nombre    1
direccion_municipio_id        1
direccion_municipio_nombre    1
direccion_isla_id             1
direccion_isla_nombre         1
telefono_1                    1
telefono_2                    1
pagina_web                    1
area_id                       3
area_nombre                   3
subarea_id                    2
subarea_nombre                2
dtype: int64

In [31]:
different_columns = distinct_counts[distinct_counts > 1].index
display(selected[different_columns])

,numero_registro,direccion_nombre_via,direccion_numero,area_id,area_nombre,subarea_id,subarea_nombre
601,ERE2022TF00115,Calle Luis Saavedra Miranda,8,6,Exclusión Social,_U,_U
602,ERE2022TF00115,Calle Luis Saavedra Miranda,8,9,Voluntariado,2,Servicios Sociales
711,ERE2024TF00037,Don Pio Coronado,118,3,Discapacidad,_U,_U


### `nif` — conclusion

- The field represents the organisation's tax identification number
  and should be stored as a string.
- There are 810 values, of which 745 are distinct.
- No null values, `_U` placeholders, blank strings, or leading/trailing
  whitespace were detected.
- All values have 9 characters; tax identifier validity has not been checked.
- `nif` does not uniquely identify a row. The highest observed frequency
  of a single NIF is 7 records.
- In this dataset, each NIF corresponds to exactly one organisation name,
  and each organisation name corresponds to exactly one NIF.
- Each registration number corresponds to exactly one NIF.
  741 NIFs have one registration number, while 4 NIFs have two distinct
  registration numbers: an observed one-to-many relationship.
- Rows should not be removed solely because they share a NIF, since they
  may contain different registrations, addresses, or activity categories.
- These findings apply to the analysed dataset.

### `nif` — conclusión

- El campo representa el número de identificación fiscal de la organización
  y debe conservarse como texto.
- Hay 810 valores, de los cuales 745 son distintos.
- No se detectaron valores nulos, marcadores `_U`, cadenas vacías
  ni espacios al principio o al final.
- Todos los valores tienen 9 caracteres; no se ha comprobado su validez fiscal.
- `nif` no identifica una fila de forma única. La frecuencia máxima observada
  de un mismo NIF es de 7 registros.
- En este conjunto de datos, cada NIF corresponde a una única denominación
  y cada denominación corresponde a un único NIF.
- Cada número de registro corresponde a un único NIF.
  741 NIF tienen un número de registro, mientras que 4 NIF tienen dos números
  de registro distintos: una relación observada de uno a varios.
- No deben eliminarse filas únicamente por compartir el mismo NIF, ya que
  pueden contener diferentes registros, direcciones o categorías de actividad.
- Estos resultados se refieren al conjunto de datos analizado.

In [58]:
col = "denominacion"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))


dtype: str
missing: 0
non-null: 810
unique: 745


0                                    Asociacion Adepsi
1                         Radio Ecca Fundacion Canaria
2                           Ciudad De San Juan De Dios
3    Civitas Centros Especializados En Personas Con...
4    Asociación De Pensionistas Jubilados Por Vejez...
5                    Asociación De Mayores San Nicolás
6    Asociación De Jubilados Y Pensionistas Párroco...
7            Asoc Cultural De La Tercera Edad Guayasen
8                   Asociación Tercera Edad Breña Alta
9                                               Aldeas
Name: denominacion, dtype: str

denominacion
Asoc Canaria Oceano De Luz                                                     7
Fundacion Ecca Social                                                          5
Asoc Domitila Hernandez Desde Tacoronte Por La Iguadad De Oportunidades        3
Cardiolatidos Servicios De Formacion Y Ayuda A Dependientes                    3
Fundacion Maroparque Fundacion Canaria                                         3
Fundacion Canaria Centro De Orientacion Familiar De Canaria                    3
Asoc Las Palmas Invita                                                         3
Asociacion Cesal                                                               3
Fundación Canaria Global Iniciativas Sociales                                  2
Asoc. Real Soc. Económica Amigos País Gran Canaria                             2
Asociacion Psicosocial Tu Vida Con Dolor                                       2
Asoc Socioeducativa Asermi                                                     2
Lares Asociacio

In [59]:
names = df["denominacion"]

In [60]:
lengths = names.str.len()
display(lengths.describe())

count    810.000000
mean      42.282716
std       19.351356
min        5.000000
25%       29.000000
50%       39.000000
75%       54.000000
max      124.000000
Name: denominacion, dtype: float64

In [73]:
display(names[(lengths < lengths.min() + 4)])

9        Aldeas
229     Fecanco
246      Adecco
269    Aciudesa
284      Asispa
490      Apaelp
534    Diaconia
675    Vitaliti
699     Confuer
760       Silva
Name: denominacion, dtype: str

In [90]:
with pd.option_context('display.max_colwidth', None):
    display(names[(lengths > lengths.max() - 28)].sort_values())

457                            Asociacion De Familiares Y Cuidadores De Mayores Dependientes Enfermos De Alzheimer Y Otras Demencia
157                             Asociacion De Mujeres Empresarias Directivas Y Cargos De Responsabilidad De La Comunidad Autonoma C
319                            Asociacion Educativa Formativa Jaslem Ayuda Y Apoyo A Las Personas Con Sindrome De Down Y Otrtas Dis
782                               Asociacion Para La Integracion Y Promocion De Las Personas Con Discapacidad Atletas Sin Fronteras
67     Asociación Colectivo Gamá - Asociación De Lesbianas Gays Trans Y Bisexuales De Canarias Por La Diversidad Sexual Y De Género
299                             Comedor Social N.s. Del Carmen Compañia Hijas De La Caridad De San Vicente De Paul Provincia España
571                               Comunidad Obra Social La Milagrosa (cía. De Las Hijas De La Caridad De S. Vicente De Paúl España)
321                              Federacion Coordinadora De Organizaciones N

In [81]:
value = names.loc[457]
print(repr(value))
print("Length:", len(value))
print("Last 40 characters:", value[-40:])

'Asociacion De Familiares Y Cuidadores De Mayores Dependientes Enfermos De Alzheimer Y Otras Demencia'
Length: 100
Last 40 characters: s Enfermos De Alzheimer Y Otras Demencia


In [82]:
profile = pd.DataFrame({"name": names, "length": lengths})

with pd.option_context("display.max_colwidth", None):
    display(profile.nsmallest(10, "length"))
    display(profile.nlargest(10, "length"))

,name,length
760,Silva,5
9,Aldeas,6
246,Adecco,6
284,Asispa,6
490,Apaelp,6
229,Fecanco,7
699,Confuer,7
269,Aciudesa,8
534,Diaconia,8
675,Vitaliti,8


,name,length
67,Asociación Colectivo Gamá - Asociación De Lesbianas Gays Trans Y Bisexuales De Canarias Por La Diversidad Sexual Y De Género,124
319,Asociacion Educativa Formativa Jaslem Ayuda Y Apoyo A Las Personas Con Sindrome De Down Y Otrtas Dis,100
457,Asociacion De Familiares Y Cuidadores De Mayores Dependientes Enfermos De Alzheimer Y Otras Demencia,100
157,Asociacion De Mujeres Empresarias Directivas Y Cargos De Responsabilidad De La Comunidad Autonoma C,99
299,Comedor Social N.s. Del Carmen Compañia Hijas De La Caridad De San Vicente De Paul Provincia España,99
316,Federacion Provincial De Empresas Del Metal Y Nuevas Teconologias De Santa Cruz De Tenerife Femete,98
321,Federacion Coordinadora De Organizaciones No Gubernamentales Para El Desarrollo (ongd) De Canarias,98
59,Fundacion Para La Formacion Integral Y Promocion Sacio Laboral Y Coop Para El Desarrollo Isonorte,97
571,Comunidad Obra Social La Milagrosa (cía. De Las Hijas De La Caridad De S. Vicente De Paúl España),97
782,Asociacion Para La Integracion Y Promocion De Las Personas Con Discapacidad Atletas Sin Fronteras,97


In [83]:
names = df["denominacion"]

mask = names != names.str.strip()
display(names[mask])

Series([], Name: denominacion, dtype: str)

In [85]:
normalized_names = (
    names
    .str.strip()
    .str.lower()
    .str.replace(r"\s+", " ", regex=True)
)
display(normalized_names)

0                                      asociacion adepsi
1                           radio ecca fundacion canaria
2                             ciudad de san juan de dios
3      civitas centros especializados en personas con...
4      asociación de pensionistas jubilados por vejez...
                             ...                        
805    asoc de personas con diversidad funcional asmi...
806             asociacion mujeres unidas del mundo-amum
807             asociacion mujeres unidas del mundo-amum
808                    fundacion canaria mexico canarias
809    asoc comité canario de representantes de perso...
Name: denominacion, Length: 810, dtype: str

In [87]:
print("Rows:", len(names))
print("unique before:", names.nunique())
print("unique after:", normalized_names.nunique())

Rows: 810
unique before: 745
unique after: 745


In [88]:
names[names != names.str.strip()]

Series([], Name: denominacion, dtype: str)

In [91]:
display(profile.loc[[157, 319, 457]])

,name,length
157,Asociacion De Mujeres Empresarias Directivas Y...,99
319,Asociacion Educativa Formativa Jaslem Ayuda Y ...,100
457,Asociacion De Familiares Y Cuidadores De Mayor...,100


In [92]:
with pd.option_context("display.max_colwidth", None):
    display(profile.loc[profile["length"] > 100])

,name,length
67,Asociación Colectivo Gamá - Asociación De Lesbianas Gays Trans Y Bisexuales De Canarias Por La Diversidad Sexual Y De Género,124


### `denominacion` — profile and data quality conclusions

- The field contains organisation names stored as strings.
- There are 810 non-null values and 745 distinct names.
- No blank strings or leading/trailing whitespace were detected.
- The highest observed frequency of a single name is 7 records.
- Name lengths range from 5 to 124 characters.
- Normalizing letter case and whitespace did not change the number
  of distinct names.
- Names at row indices 157, 319, and 457 have apparently incomplete
  endings and lengths of 99, 100, and 100 characters, respectively.
  These endings were also confirmed in the source JSON.
- This suggests possible truncation in the source. Its cause is unknown;
  a universal 100-character limit cannot be inferred because longer
  names also exist.
- A possible spelling error (`Otrtas`) was observed in row 319.
- In this dataset, `denominacion` and `nif` have a one-to-one relationship.
- The relationship from `denominacion` to `numero_registro` is one-to-many:
  741 names correspond to one registration number and 4 names to two.
  Each registration number corresponds to exactly one name.
- Organisation names do not uniquely identify rows. Repeated names
  should not be removed without examining the other fields.
- Preserve source values; flag suspected truncation and spelling errors
  for review rather than correcting them without supporting evidence.
- These findings apply to the analysed dataset.

### `denominacion` — conclusiones del perfil y de la calidad de los datos

- El campo contiene nombres de organizaciones almacenados como texto.
- Hay 810 valores no nulos y 745 denominaciones distintas.
- No se detectaron cadenas vacías ni espacios al principio o al final.
- La frecuencia máxima observada de una misma denominación es de 7 registros.
- Las longitudes varían entre 5 y 124 caracteres.
- La normalización de mayúsculas/minúsculas y espacios no modificó
  el número de denominaciones distintas.
- Las denominaciones de las filas con índices 157, 319 y 457 presentan
  finales aparentemente incompletos y longitudes de 99, 100 y 100 caracteres,
  respectivamente. Estos finales también se confirmaron en el JSON de origen.
- Esto sugiere un posible truncamiento en el origen. Se desconoce su causa;
  no puede deducirse un límite general de 100 caracteres, ya que también
  existen denominaciones más largas.
- Se observó una posible errata (`Otrtas`) en la fila 319.
- En este conjunto de datos, `denominacion` y `nif` tienen una relación
  de uno a uno.
- La relación de `denominacion` a `numero_registro` es de uno a varios:
  741 denominaciones corresponden a un número de registro y 4 a dos.
  Cada número de registro corresponde a una única denominación.
- Las denominaciones no identifican filas de forma única. No deben eliminarse
  nombres repetidos sin examinar los demás campos.
- Deben conservarse los valores originales y señalarse los posibles
  truncamientos y erratas para su revisión, sin corregirlos sin evidencia.
- Estos resultados se refieren al conjunto de datos analizado.

In [94]:
col = "numero_registro"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))


dtype: str
missing: 0
non-null: 810
unique: 749


0    ERE1986CA00001
1    ERE1986CA00002
2    ERE1987CA00001
3    ERE1987CA00002
4    ERE1987CA00003
5    ERE1987CA00004
6    ERE1987CA00005
7    ERE1987CA00007
8    ERE1988CA00001
9    ERE1989CA00001
Name: numero_registro, dtype: str

numero_registro
ERE2023TF00106    7
ERE2023GC00013    5
ERE2021TF00125    3
ERE2022TF00113    3
ERE2022TF00114    3
ERE2024TF00104    3
ERE2021TF00082    2
ERE2021TF00117    2
ERE2021TF00118    2
ERE2021TF00119    2
ERE2021TF00122    2
ERE2021TF00126    2
ERE2022TF00017    2
ERE2022TF00018    2
ERE2022TF00019    2
ERE2022TF00033    2
ERE2022TF00036    2
ERE2022TF00111    2
ERE2022TF00115    2
ERE2023GC00001    2
Name: count, dtype: int64

In [95]:
registrations = df["numero_registro"]

In [96]:
display(registrations.str.len().value_counts())

numero_registro
14    810
Name: count, dtype: int64

In [97]:
print("Placeholder _U:", registrations.eq("_U").sum())
print("Blank values:", registrations.str.strip().eq("").sum())
print("Leading/trailing whitespace:", registrations.ne(registrations.str.strip()).sum())

Placeholder _U: 0
Blank values: 0
Leading/trailing whitespace: 0


In [98]:
pattern = r"[A-Z]{3}[0-9]{4}[A-Z]{2}[0-9]{5}"
matches = registrations.str.fullmatch(pattern, na=False)
display(matches.value_counts())

numero_registro
True    810
Name: count, dtype: int64

### `numero_registro` — conclusions

- The field contains registration numbers stored as strings.
- There are 810 non-null values and 749 distinct registration numbers.
- No `_U` placeholders, blank strings, or leading/trailing whitespace
  were detected.
- All values contain 14 characters and match the observed format:
  three uppercase Latin letters, four digits, two uppercase Latin letters,
  and five digits. This confirms format consistency, not registration validity.
- The highest observed frequency of a single registration number is 7 records.
- `numero_registro` does not uniquely identify a row.

#### Relationships with other fields

- Each `numero_registro` corresponds to exactly one `nif`
  and one `denominacion`.
- The relationship from `nif` to `numero_registro` is one-to-many:
  741 NIFs have one registration number and 4 NIFs have two distinct numbers.
- Likewise, 741 distinct organisation names correspond to one registration
  number and 4 names correspond to two.
- Records sharing a registration number may contain different activity
  categories (`area_id`, `area_nombre`, `subarea_id`, `subarea_nombre`),
  as observed in the inspected examples.
- Repeated registration numbers alone are not sufficient grounds
  for removing rows.
- Preserve the field as a string. Its role in the global schema will depend
  on whether the target entity represents an organisation or a registration.
- These findings apply to the analysed dataset.

### `numero_registro` — conclusiones

- El campo contiene números de registro almacenados como texto.
- Hay 810 valores no nulos y 749 números de registro distintos.
- No se detectaron marcadores `_U`, cadenas vacías ni espacios
  al principio o al final.
- Todos los valores tienen 14 caracteres y cumplen el formato observado:
  tres letras latinas mayúsculas, cuatro dígitos, dos letras latinas
  mayúsculas y cinco dígitos. Esto confirma la uniformidad del formato,
  no la validez del registro.
- La frecuencia máxima observada de un mismo número de registro
  es de 7 registros.
- `numero_registro` no identifica una fila de forma única.

#### Relaciones con otros campos

- Cada `numero_registro` corresponde a un único `nif`
  y a una única `denominacion`.
- La relación de `nif` a `numero_registro` es de uno a varios:
  741 NIF tienen un número de registro y 4 NIF tienen dos números distintos.
- Asimismo, 741 denominaciones distintas corresponden a un número
  de registro y 4 denominaciones corresponden a dos.
- Las filas que comparten un número de registro pueden contener diferentes
  categorías de actividad (`area_id`, `area_nombre`, `subarea_id`,
  `subarea_nombre`), como se observó en los ejemplos examinados.
- La repetición del número de registro no basta para justificar
  la eliminación de filas.
- El campo debe conservarse como texto. Su función en el esquema global
  dependerá de si la entidad de destino representa una organización
  o una inscripción registral.
- Estos resultados se refieren al conjunto de datos analizado.

In [100]:
col = "direccion_tipo_via"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
print("Blank values:", df[col].str.strip().eq("").sum())
print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))


dtype: str
missing: 0
non-null: 810
unique: 23
Placeholder _U: 0
Blank values: 49
Leading/trailing whitespace: 49


0    Calle
1     Avda
2     Ctra
3    Calle
4    Calle
5    Calle
6    Calle
7    Calle
8    Calle
9    Calle
Name: direccion_tipo_via, dtype: str

direccion_tipo_via
Calle    583
Avda      71
          49
Ctra      27
Plaza     25
Cmno      16
Urb        8
Rbla       6
Carre      4
Lugar      3
Paseo      3
Psaje      3
Trva       2
Pque       1
Zona       1
Custa      1
Muell      1
Sbida      1
Barda      1
Senda      1
Name: count, dtype: int64

In [101]:
blank_mask = df[col].str.strip().eq("")
whitespace_mask = df[col].ne(df[col].str.strip())

print(blank_mask.equals(whitespace_mask))

True


In [102]:
display(df.loc[blank_mask, col].map(repr).value_counts())

direccion_tipo_via
' '    49
Name: count, dtype: int64

In [103]:
display(
    df.loc[
        blank_mask,
        ["direccion_tipo_via", "direccion_nombre_via", "direccion_numero"]
    ].head(15)
)

,direccion_tipo_via,direccion_nombre_via,direccion_numero
10,,Urb San Diego,5
14,,Perez Del Toro,14
26,,Avenida De Escaleritas,66
41,,C/ Dacil Vilar Borges Locales 35 Y 36,35
59,,El Cardal,6
63,,C/ Doctor Verneau,1
88,,Calle Cordoba,15
117,,C/antonio Manchado Viglietti,1
131,,Miguel Primo De Rivera,6
139,,Avenida Primero De Mayo,21


In [105]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 810 entries, 0 to 809
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype
---  ------                      --------------  -----
 0   nif                         810 non-null    str  
 1   denominacion                810 non-null    str  
 2   numero_registro             810 non-null    str  
 3   direccion_tipo_via          810 non-null    str  
 4   direccion_nombre_via        810 non-null    str  
 5   direccion_numero            810 non-null    str  
 6   direccion_complemento       810 non-null    str  
 7   direccion_provincia_id      810 non-null    str  
 8   direccion_provincia_nombre  810 non-null    str  
 9   direccion_municipio_id      810 non-null    int64
 10  direccion_municipio_nombre  810 non-null    str  
 11  direccion_isla_id           810 non-null    str  
 12  direccion_isla_nombre       810 non-null    str  
 13  telefono_1                  810 non-null    str  
 14  telefono_2           

In [108]:
display(
    df.loc[
        :,
        ##blank_mask,
        ["direccion_tipo_via", "direccion_nombre_via", "direccion_numero"]
    ].head(15)
)

,direccion_tipo_via,direccion_nombre_via,direccion_numero
0,Calle,Lomo La Plana,28
1,Avda,Avenida De Escaleritas 64,64
2,Ctra,Carretera Ciudad San Juan De Dios,1
3,Calle,La Calzada,88
4,Calle,Los Martínez De Escobar,16
5,Calle,Garagonay - Sardina Del Sur,1
6,Calle,Mirasol,52
7,Calle,Calle Cirilo Moreno,8
8,Calle,Adelfas,8
9,Calle,C/ Escultor Estevez,17


In [109]:
col = "direccion_nombre_via"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
print("Blank values:", df[col].str.strip().eq("").sum())
print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))

dtype: str
missing: 0
non-null: 810
unique: 685
Placeholder _U: 0
Blank values: 0
Leading/trailing whitespace: 0


0                        Lomo La Plana
1            Avenida De Escaleritas 64
2    Carretera Ciudad San Juan De Dios
3                           La Calzada
4              Los Martínez De Escobar
5          Garagonay - Sardina Del Sur
6                              Mirasol
7                  Calle Cirilo Moreno
8                              Adelfas
9                  C/ Escultor Estevez
Name: direccion_nombre_via, dtype: str

direccion_nombre_via
Profesor Lozano               7
Leon Y Castillo               5
Doctor Chill                  5
Primero De Mayo               4
Juan Rumeu García             4
Viera Y Clavijo               4
La Constitución               4
Triana                        3
León Y Castillo               3
Doctor Chil                   3
Antonio Manchado Viglietti    3
Bravo Murillo                 3
Calvo Sotelo                  3
La Resbala                    3
Los Almendros                 3
Adelantado                    3
Batalla Del Ebro              3
La Cuesta                     3
Siena                         3
La Calzada                    2
Name: count, dtype: int64

In [110]:
address_profile = pd.DataFrame({
    "address": df[col],
    "length": df[col].str.len()
})

with pd.option_context("display.max_colwidth", None):
    display(address_profile.nsmallest(10, "length"))
    display(address_profile.nlargest(10, "length"))

,address,length
194,D,1
123,Real,4
142,León,4
145,Goya,4
316,Mazo,4
339,Maya,4
463,Maya,4
492,Hero,4
582,Null,4
735,Tano,4


,address,length
25,Verónica Las Palmas De Gc (a Efectos De Correspondencia) C/ Córdoba 29 Bis,76
453,Subida Igueste De Candelariaesquina Ctra. General Del Sur C-28,62
373,Tirajana - Edif. Mercurio Torre 11 - Playa Del Inglés,53
232,Principes De España Granadilla De Abona - El Médano,51
231,Plaza De La Candelaria 28 Edificio Olimpo 3º 304,48
190,Nave En Parcela 124 - Mercalaspalmas - Marzagan,47
216,Puerto Del Carmen Urbanización Virgen Del Rocio,47
218,. C/. Universo Nº20. Gracia. La Laguna Canarias,47
438,Plaza De La Real Sociedad Econ. Amigos Del País,47
439,Plaza De La Real Sociedad Econ. Amigos Del País,47


In [111]:
mask = df[col].str.strip().str.lower().eq("null")

display(
    df.loc[
        mask,
        ["direccion_tipo_via", "direccion_nombre_via", "direccion_numero"]
    ]
)

,direccion_tipo_via,direccion_nombre_via,direccion_numero
582,Calle,Null,7


In [112]:
col = "direccion_numero"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
print("Blank values:", df[col].str.strip().eq("").sum())
print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))

dtype: str
missing: 0
non-null: 810
unique: 133
Placeholder _U: 0
Blank values: 0
Leading/trailing whitespace: 0


0    28
1    64
2     1
3    88
4    16
5     1
6    52
7     8
8     8
9    17
Name: direccion_numero, dtype: str

direccion_numero
S/n    57
1      48
2      38
5      34
15     32
4      29
12     21
9      21
8      20
7      20
3      20
28     19
11     18
10     17
17     16
14     15
21     15
29     15
6      14
19     13
Name: count, dtype: int64

In [113]:
only_digits = df[col].str.fullmatch(r"[0-9]+", na=False)
display(df.loc[~only_digits, col].value_counts(dropna=False))

direccion_numero
S/n      57
S/n.      2
37 I      1
26a       1
31bis     1
3 Y 5     1
A         1
34-36     1
2 Y 3     1
115/7     1
9-11      1
3 Baj     1
8-A       1
148vv     1
29b       1
Name: count, dtype: int64

In [114]:
col = "direccion_complemento"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
print("Blank values:", df[col].str.strip().eq("").sum())
print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))

dtype: str
missing: 0
non-null: 810
unique: 180
Placeholder _U: 488
Blank values: 0
Leading/trailing whitespace: 70


0         _U
1          1
2         _U
3         _U
4      Izq A
5         _U
6         _U
7    Bj Izq.
8         _U
9         12
Name: direccion_complemento, dtype: str

direccion_complemento
_U         488
Baj         30
1º          13
1           11
1 B          7
1 A          7
1 Izqui      7
2            6
 B           5
3            5
2 D          5
Bj           5
4 A          4
0            4
 2           4
8            4
5            3
2 B          3
 A           3
2º           3
Name: count, dtype: int64

In [115]:
display(df[col].value_counts(dropna=False))

direccion_complemento
_U         488
Baj         30
1º          13
1           11
1 B          7
          ... 
0 1 4        1
 Izd bj      1
2 2          1
 35          1
  Bajo       1
Name: count, Length: 180, dtype: int64

In [116]:
with pd.option_context("display.max_rows", None):
    display(df[col].value_counts(dropna=False))

direccion_complemento
_U             488
Baj             30
1º              13
1               11
1 B              7
1 A              7
1 Izqui          7
2                6
 B               5
3                5
2 D              5
Bj               5
4 A              4
0                4
 2               4
8                4
5                3
2 B              3
 A               3
2º               3
7 C              3
1 1              3
1 D              3
1 Izq            3
 25              3
 12              2
 10              2
  1              2
4º               2
4 Izq.           2
2 0              2
 8               2
2 C              2
3 C              2
 4               2
2 A              2
5 B              2
3 B              2
4 Ofi            2
5 D              2
1º D             2
B J              2
 18b             2
0 A              2
 ba              2
3 A              2
0 23             2
Baj  Bis         2
Izq A            1
Bj Izq.          1
1ºi G            1
 G       

In [117]:
print("Before stripping whitespace:", df[col].nunique())
print("After stripping whitespace:", df[col].str.strip().nunique())

Before stripping whitespace: 180
After stripping whitespace: 166


### Street address fields — conclusions

#### `direccion_tipo_via`

- String field with 23 distinct values.
- No null values or `_U` placeholders were detected.
- 49 values are blank after trimming whitespace.
- In inspected records, `direccion_nombre_via` may include or omit the street
  type both when `direccion_tipo_via` is populated and when it is blank.
  The separation of street type and street name is therefore inconsistent.

#### `direccion_nombre_via`

- Contains 810 non-null string values and 685 distinct values.
- No `_U` placeholders, blank strings, or leading/trailing whitespace
  were detected.
- One record contains the literal string `"Null"`, a suspected missing-value
  marker that pandas does not recognize as null.
- Observed lengths range from 1 to 76 characters.
- The field may contain street types, building numbers, building details,
  locality names, and correspondence instructions in addition to street names.
- Internal whitespace inconsistencies and spelling variants were observed.
- The single-character value `"D"` requires contextual review.

#### `direccion_numero`

- Contains 810 non-null string values and 133 distinct values.
- No `_U` placeholders, blank strings, or leading/trailing whitespace
  were detected.
- Values include digits, letters, ranges, multiple numbers, and additional
  address information. The field should therefore remain a string.
- `S/n` occurs 57 times and `S/n.` twice; these conventionally mean
  “without a number”.
- Values that do not consist solely of digits are not necessarily errors.

#### `direccion_complemento`

- Contains 810 non-null string values and 180 distinct values, including `_U`.
- No blank strings were detected.
- `_U` occurs 488 times (approximately 60.2% of records); its precise meaning
  is unconfirmed.
- 70 values contain leading/trailing whitespace.
- Trimming whitespace reduces the number of distinct values from 180 to 166.
- Values include floor and unit details, varied abbreviations, and repeated
  internal spaces.
- `- - - -` is another possible missing-value marker requiring confirmation.

#### Implications for later integration

- Address components are not consistently separated across these fields.
- Preserve the original values and document these issues before mapping
  to `Entidad.direccion`.
- Transformation rules will be defined at the integration stage.

### Campos de dirección — conclusiones

#### `direccion_tipo_via`

- Campo de texto con 23 valores distintos.
- No se detectaron valores nulos ni marcadores `_U`.
- 49 valores quedan vacíos al eliminar los espacios iniciales y finales.
- En los registros examinados, `direccion_nombre_via` puede incluir u omitir
  el tipo de vía tanto cuando `direccion_tipo_via` está informado como cuando
  está vacío. La separación entre tipo y nombre de vía es inconsistente.

#### `direccion_nombre_via`

- Contiene 810 valores de texto no nulos y 685 valores distintos.
- No se detectaron marcadores `_U`, cadenas vacías ni espacios iniciales
  o finales.
- Un registro contiene el texto literal `"Null"`, un posible marcador
  de información ausente que pandas no reconoce como valor nulo.
- Las longitudes observadas varían entre 1 y 76 caracteres.
- Además del nombre de la vía, el campo puede incluir tipos de vía, números,
  detalles de edificios, localidades e instrucciones de correspondencia.
- Se observaron inconsistencias en los espacios internos y variantes
  de escritura.
- El valor de un solo carácter `"D"` requiere revisión en su contexto.

#### `direccion_numero`

- Contiene 810 valores de texto no nulos y 133 valores distintos.
- No se detectaron marcadores `_U`, cadenas vacías ni espacios iniciales
  o finales.
- Los valores incluyen dígitos, letras, intervalos, varios números
  e información adicional de dirección. Debe conservarse como texto.
- `S/n` aparece 57 veces y `S/n.` dos veces; convencionalmente significan
  “sin número”.
- Los valores que no contienen exclusivamente dígitos no son necesariamente
  errores.

#### `direccion_complemento`

- Contiene 810 valores de texto no nulos y 180 valores distintos, incluido `_U`.
- No se detectaron cadenas vacías.
- `_U` aparece 488 veces (aproximadamente el 60,2% de los registros);
  su significado exacto no está confirmado.
- 70 valores contienen espacios iniciales o finales.
- Al eliminarlos, el número de valores distintos disminuye de 180 a 166.
- Los valores incluyen detalles de plantas y unidades, abreviaturas variadas
  y espacios internos repetidos.
- `- - - -` es otro posible marcador de información ausente que requiere
  confirmación.

#### Implicaciones para la integración posterior

- Los componentes de la dirección no están separados de forma uniforme.
- Deben conservarse los valores originales y documentarse estas incidencias
  antes de asociarlos con `Entidad.direccion`.
- Las reglas de transformación se definirán en la etapa de integración.

In [119]:
col = "direccion_provincia_id"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
print("Blank values:", df[col].str.strip().eq("").sum())
print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))

dtype: str
missing: 0
non-null: 810
unique: 13
Placeholder _U: 8
Blank values: 0
Leading/trailing whitespace: 0


0    35
1    35
2    35
3    35
4    35
5    35
6    35
7    35
8    38
9    38
Name: direccion_provincia_id, dtype: str

direccion_provincia_id
38    396
35    374
28     17
_U      8
36      4
30      2
46      2
11      2
49      1
50      1
40      1
47      1
06      1
Name: count, dtype: int64

In [120]:
col = "direccion_provincia_nombre"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
print("Blank values:", df[col].str.strip().eq("").sum())
print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))

dtype: str
missing: 0
non-null: 810
unique: 13
Placeholder _U: 8
Blank values: 0
Leading/trailing whitespace: 0


0                Las Palmas
1                Las Palmas
2                Las Palmas
3                Las Palmas
4                Las Palmas
5                Las Palmas
6                Las Palmas
7                Las Palmas
8    Santa Cruz de Tenerife
9    Santa Cruz de Tenerife
Name: direccion_provincia_nombre, dtype: str

direccion_provincia_nombre
Santa Cruz de Tenerife    396
Las Palmas                374
Madrid                     17
_U                          8
Pontevedra                  4
Murcia                      2
Valencia                    2
Cadiz                       2
Zamora                      1
Zaragoza                    1
Segovia                     1
Valladolid                  1
Badajoz                     1
Name: count, dtype: int64

In [121]:
province_columns = [
    "direccion_provincia_id",
    "direccion_provincia_nombre"
]

display(
    df[province_columns]
    .drop_duplicates()
    .sort_values("direccion_provincia_id")
)

,direccion_provincia_id,direccion_provincia_nombre
778,06,Badajoz
419,11,Cadiz
44,28,Madrid
39,30,Murcia
0,35,Las Palmas
622,36,Pontevedra
8,38,Santa Cruz de Tenerife
352,40,Segovia
383,46,Valencia
655,47,Valladolid


In [122]:
df.groupby("direccion_provincia_id")[
    "direccion_provincia_nombre"
].nunique().value_counts()

direccion_provincia_nombre
1    13
Name: count, dtype: int64

In [123]:
df.groupby("direccion_provincia_nombre")[
    "direccion_provincia_id"
].nunique().value_counts()

direccion_provincia_id
1    13
Name: count, dtype: int64

### Province fields — conclusions

- `direccion_provincia_id` and `direccion_provincia_nombre` are string fields.
- Both contain 810 non-null values, with no blank strings or leading/trailing
  whitespace.
- Each field has 13 distinct values: 12 province codes/names and the marker `_U`.
- The pair `_U → _U` occurs in 8 records and represents potentially missing
  province information; its exact meaning remains unconfirmed.
- In this dataset, `direccion_provincia_id` and `direccion_provincia_nombre`
  have a one-to-one relationship: each code corresponds to exactly one name,
  and each name corresponds to exactly one code.
- Province codes should remain strings to preserve leading zeros, such as `"06"`.
- These fields are candidates for mapping to `Provincia.codigo`
  and `Provincia.nombre` in the global schema.
- These findings apply to the analysed dataset.

### Campos de provincia — conclusiones

- `direccion_provincia_id` y `direccion_provincia_nombre` son campos de texto.
- Ambos contienen 810 valores no nulos, sin cadenas vacías ni espacios
  al principio o al final.
- Cada campo tiene 13 valores distintos: 12 códigos/nombres de provincias
  y el marcador `_U`.
- La pareja `_U → _U` aparece en 8 registros y representa información
  provincial posiblemente ausente; su significado exacto no está confirmado.
- En este conjunto de datos, `direccion_provincia_id` y
  `direccion_provincia_nombre` tienen una relación de uno a uno:
  cada código corresponde a un único nombre y cada nombre a un único código.
- Los códigos deben conservarse como texto para mantener los ceros iniciales,
  como en `"06"`.
- Estos campos son candidatos para asociarse con `Provincia.codigo`
  y `Provincia.nombre` en el esquema global.
- Estos resultados se refieren al conjunto de datos analizado.

In [127]:
col = "direccion_municipio_id"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
##print("Blank values:", df[col].str.strip().eq("").sum())
##print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))

dtype: int64
missing: 0
non-null: 810
unique: 90
Placeholder _U: 0


0    35001
1    35001
2    35001
3    35001
4    35001
5    35190
6    35001
7    35001
8    38080
9    38230
Name: direccion_municipio_id, dtype: int64

direccion_municipio_id
35001    229
38001    126
38230     91
35140     22
38270     20
38170     19
35230     19
28001     17
35030     17
35160     15
38240     11
38110     10
38220      9
38430      9
38290      9
35190      7
38180      7
38250      7
38080      6
35120      6
Name: count, dtype: int64

In [128]:
col = "direccion_municipio_nombre"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
print("Blank values:", df[col].str.strip().eq("").sum())
print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))

dtype: str
missing: 0
non-null: 810
unique: 84
Placeholder _U: 8
Blank values: 0
Leading/trailing whitespace: 0


0    Palmas de Gran Canaria (Las)
1    Palmas de Gran Canaria (Las)
2    Palmas de Gran Canaria (Las)
3    Palmas de Gran Canaria (Las)
4    Palmas de Gran Canaria (Las)
5         Santa Lucía de Tirajana
6    Palmas de Gran Canaria (Las)
7    Palmas de Gran Canaria (Las)
8                      Breña Alta
9                     Laguna (La)
Name: direccion_municipio_nombre, dtype: str

direccion_municipio_nombre
Palmas de Gran Canaria (Las)    229
Santa Cruz de Tenerife          126
Laguna (La)                      91
Puerto del Rosario               22
Orotava (La)                     20
Granadilla de Abona              19
Telde                            19
Madrid                           17
Arrecife                         17
San Bartolomé de Tirajana        15
Llanos de Aridane (Los)          11
Candelaria                       10
Icod de los Vinos                 9
Tacoronte                         9
Puerto de la Cruz                 9
_U                                8
Santa Lucía de Tirajana           7
Guancha (La)                      7
Matanza de Acentejo (La)          7
Breña Alta                        6
Name: count, dtype: int64

In [129]:
municipio_columns = [
    "direccion_municipio_id",
    "direccion_municipio_nombre"
]

display(
    df[municipio_columns]
    .drop_duplicates()
    .sort_values("direccion_municipio_id")
)

,direccion_municipio_id,direccion_municipio_nombre
778,6001,Badajoz
419,11080,Barrios (Los)
689,11210,Linea de la Concepcion (La)
44,28001,Madrid
39,30001,Murcia
...,...,...
352,40144,Campo de San Pedro
383,46001,Valencia
655,47001,Valladolid
168,49001,Zamora


In [130]:
df.groupby("direccion_municipio_id")[
    "direccion_municipio_nombre"
].nunique().value_counts()

direccion_municipio_nombre
1    90
Name: count, dtype: int64

In [131]:
df.groupby("direccion_municipio_nombre")[
    "direccion_municipio_id"
].nunique().value_counts()

direccion_municipio_id
1    83
7     1
Name: count, dtype: int64

In [132]:
display(
    df.loc[
        df["direccion_municipio_nombre"].eq("_U"),
        municipio_columns
    ]
)

,direccion_municipio_id,direccion_municipio_nombre
162,35003,_U
272,38111,_U
285,35009,_U
294,35015,_U
296,38008,_U
526,35007,_U
579,35003,_U
717,35004,_U


In [133]:
dir_u = df.loc[df["direccion_municipio_nombre"].eq("_U"), "direccion_municipio_id"]
display(df.loc[df["direccion_municipio_id"].isin(dir_u) & df["direccion_municipio_nombre"].ne("_U"), municipio_columns])

,direccion_municipio_id,direccion_municipio_nombre


In [134]:
df.groupby("direccion_municipio_id")[
    "direccion_provincia_id"
].nunique().value_counts()

direccion_provincia_id
1    90
Name: count, dtype: int64

In [135]:
display(
    df.loc[
        df["direccion_provincia_id"].eq("_U"),
        ["direccion_municipio_id",
         "direccion_municipio_nombre",
         "direccion_provincia_id"]
    ].drop_duplicates()
)

,direccion_municipio_id,direccion_municipio_nombre,direccion_provincia_id
162,35003,_U,_U
272,38111,_U,_U
285,35009,_U,_U
294,35015,_U,_U
296,38008,_U,_U
526,35007,_U,_U
717,35004,_U,_U


### Municipality fields — conclusions

- `direccion_municipio_id` is stored as `int64`; it represents an identifier,
  not a quantity.
- It contains 810 non-null values and 90 distinct codes.
- `direccion_municipio_nombre` contains 810 non-null string values and
  84 distinct values, including `_U`.
- No blank names or leading/trailing whitespace were detected.
- `_U` occurs in 8 records covering 7 distinct municipality codes.
- No names other than `_U` were found elsewhere in this dataset for those
  7 codes.

#### Relationships between fields

- Each `direccion_municipio_id` corresponds to exactly one value of
  `direccion_municipio_nombre`.
- Each of the 83 names other than `_U` corresponds to exactly one code.
  Thus, code and name have a one-to-one relationship among named municipalities.
- `_U` corresponds to 7 codes and must not be treated as the name of
  a single municipality.
- Each municipality code corresponds to exactly one value of
  `direccion_provincia_id`. However, the province is also `_U` in the
  8 records with an unknown municipality name.
- Therefore, a known province cannot be assigned to those records using
  this dataset alone.

#### Implications for later integration

- These fields are candidates for `Localidad.codigo` and `Localidad.nombre`.
- Preserve municipality codes, including those whose names are unknown.
- Converting codes to strings may be appropriate for the global schema;
  code validity and any required leading zeros have not yet been verified.
- Records containing `_U` require explicit handling during integration.
- These findings apply to the analysed dataset.

### Campos de municipio — conclusiones

- `direccion_municipio_id` está almacenado como `int64`; representa
  un identificador, no una cantidad.
- Contiene 810 valores no nulos y 90 códigos distintos.
- `direccion_municipio_nombre` contiene 810 valores de texto no nulos
  y 84 valores distintos, incluido `_U`.
- No se detectaron nombres vacíos ni espacios al principio o al final.
- `_U` aparece en 8 registros correspondientes a 7 códigos municipales distintos.
- Para esos 7 códigos no se encontraron nombres diferentes de `_U`
  en ninguna otra fila del conjunto de datos.

#### Relaciones entre los campos

- Cada `direccion_municipio_id` corresponde a un único valor de
  `direccion_municipio_nombre`.
- Cada una de las 83 denominaciones diferentes de `_U` corresponde
  a un único código. Por tanto, existe una relación de uno a uno
  entre código y nombre para los municipios con nombre informado.
- `_U` corresponde a 7 códigos y no debe interpretarse como el nombre
  de un único municipio.
- Cada código municipal corresponde a un único valor de
  `direccion_provincia_id`. Sin embargo, la provincia también es `_U`
  en los 8 registros cuyo nombre municipal es desconocido.
- Por tanto, no puede asignarse una provincia conocida a esos registros
  utilizando únicamente este conjunto de datos.

#### Implicaciones para la integración posterior

- Estos campos son candidatos para `Localidad.codigo` y `Localidad.nombre`.
- Deben conservarse los códigos municipales, incluso cuando falta el nombre.
- Puede ser apropiado convertir los códigos a texto para el esquema global;
  todavía no se han verificado su validez ni los posibles ceros iniciales.
- Los registros con `_U` requieren un tratamiento explícito durante
  la integración.
- Estos resultados se refieren al conjunto de datos analizado.

In [136]:
col = "direccion_isla_id"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
print("Blank values:", df[col].str.strip().eq("").sum())
print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))

dtype: str
missing: 0
non-null: 810
unique: 8
Placeholder _U: 40
Blank values: 0
Leading/trailing whitespace: 0


0    GC
1    GC
2    GC
3    GC
4    GC
5    GC
6    GC
7    GC
8    LP
9    TF
Name: direccion_isla_id, dtype: str

direccion_isla_id
TF    349
GC    310
_U     40
FU     36
LP     34
LA     28
HI      7
GO      6
Name: count, dtype: int64

In [137]:
col = "direccion_isla_nombre"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
print("Blank values:", df[col].str.strip().eq("").sum())
print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))

dtype: str
missing: 0
non-null: 810
unique: 8
Placeholder _U: 40
Blank values: 0
Leading/trailing whitespace: 0


0    Gran Canaria
1    Gran Canaria
2    Gran Canaria
3    Gran Canaria
4    Gran Canaria
5    Gran Canaria
6    Gran Canaria
7    Gran Canaria
8        La Palma
9        Tenerife
Name: direccion_isla_nombre, dtype: str

direccion_isla_nombre
Tenerife         349
Gran Canaria     310
_U                40
Fuerteventura     36
La Palma          34
Lanzarote         28
El Hierro          7
La Gomera          6
Name: count, dtype: int64

In [138]:
island_columns = ["direccion_isla_id", "direccion_isla_nombre"]

display(
    df[island_columns]
    .drop_duplicates()
    .sort_values("direccion_isla_id")
)

,direccion_isla_id,direccion_isla_nombre
56,FU,Fuerteventura
0,GC,Gran Canaria
37,GO,La Gomera
32,HI,El Hierro
130,LA,Lanzarote
8,LP,La Palma
9,TF,Tenerife
39,_U,_U


In [144]:
df.groupby("direccion_isla_id")[
    "direccion_isla_nombre"
].nunique().value_counts()

direccion_isla_nombre
1    8
Name: count, dtype: int64

In [143]:
df.groupby("direccion_isla_nombre")[
    "direccion_isla_id"
].nunique().value_counts()

direccion_isla_id
1    8
Name: count, dtype: int64

In [145]:
display(
    df.loc[
        df["direccion_isla_id"].eq("_U"),
        ["direccion_provincia_id", "direccion_provincia_nombre"]
    ].value_counts()
)

direccion_provincia_id  direccion_provincia_nombre
28                      Madrid                        17
_U                      _U                             8
36                      Pontevedra                     4
30                      Murcia                         2
46                      Valencia                       2
11                      Cadiz                          2
49                      Zamora                         1
50                      Zaragoza                       1
40                      Segovia                        1
47                      Valladolid                     1
06                      Badajoz                        1
Name: count, dtype: int64

### Island fields — conclusions

- `direccion_isla_id` and `direccion_isla_nombre` are string fields.
- Both contain 810 non-null values, with no blank strings or leading/trailing
  whitespace.
- Each field contains 8 distinct values: 7 island codes/names and `_U`.
- Code and name have a one-to-one relationship in this dataset,
  including the marker pair `_U → _U`.
- `_U` occurs in 40 records: 32 have a named mainland province,
  while 8 also have `_U` in the province fields.
- This suggests that `_U` may represent either “not applicable” or
  “unknown”, depending on context; its exact meaning remains unconfirmed.
- `_U` must not be treated as an actual island.
- Preserve the island information for later integration. Its role in
  the global schema will be decided during mapping.
- These findings apply to the analysed dataset.

### Campos de isla — conclusiones

- `direccion_isla_id` y `direccion_isla_nombre` son campos de texto.
- Ambos contienen 810 valores no nulos, sin cadenas vacías ni espacios
  al principio o al final.
- Cada campo contiene 8 valores distintos: 7 códigos/nombres de islas y `_U`.
- Existe una relación de uno a uno entre código y nombre en este conjunto
  de datos, incluida la pareja de marcadores `_U → _U`.
- `_U` aparece en 40 registros: 32 tienen una provincia peninsular informada
  y 8 también contienen `_U` en los campos de provincia.
- Esto sugiere que `_U` puede representar “no aplicable” o “desconocido”,
  según el contexto; su significado exacto no está confirmado.
- `_U` no debe interpretarse como una isla real.
- Debe conservarse la información de isla para la integración posterior.
  Su función en el esquema global se decidirá durante el mapeo.
- Estos resultados se refieren al conjunto de datos analizado.

In [146]:
col = "telefono_1"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
print("Blank values:", df[col].str.strip().eq("").sum())
print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))

dtype: str
missing: 0
non-null: 810
unique: 728
Placeholder _U: 0
Blank values: 0
Leading/trailing whitespace: 0


0    (34)928414484
1    (34)928289168
2    (34)928339080
3    (34)928350138
4    (34)928263023
5    (34)928755654
6    (34)928380311
7    (34)928385140
8    (34)922437523
9    (34)666835467
Name: telefono_1, dtype: str

telefono_1
(34)608027061    7
(34)928930058    5
(34)628747191    4
(34)922573289    3
(34)828787880    3
(34)922417782    3
(34)928200006    3
(34)629575627    3
(34)629027811    3
(34)651982601    2
(34)922463646    2
(34)922815921    2
(34)922882917    2
(34)928261612    2
(34)928335405    2
(34)922255289    2
(34)928313398    2
(34)922578121    2
(34)607934308    2
(34)922496118    2
Name: count, dtype: int64

In [147]:
col = "telefono_2"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
print("Blank values:", df[col].str.strip().eq("").sum())
print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))

dtype: str
missing: 0
non-null: 810
unique: 171
Placeholder _U: 625
Blank values: 0
Leading/trailing whitespace: 0


0               _U
1               _U
2               _U
3               _U
4               _U
5               _U
6               _U
7               _U
8    (34)655221013
9               _U
Name: telefono_2, dtype: str

telefono_2
_U               625
(34)628747191      3
(34)828787881      3
(34)628019847      3
(34)928200006      3
(34)298641241      2
(34)617442956      2
(34)686615991      2
(34)664191772      2
(34)659719369      2
(34)914686941      2
(34)670663611      2
(34)655221013      1
(34)922248361      1
(34)922333872      1
(34)633758067      1
(34)630821359      1
(34)922141074      1
(34)617689889      1
(34)822664419      1
Name: count, dtype: int64

In [149]:
for col in ["telefono_1", "telefono_2"]:
    print(col)
    display(df[col].str.len().value_counts())

telefono_1


telefono_1
13    803
12      2
14      2
16      1
5       1
11      1
Name: count, dtype: int64

telefono_2


telefono_2
2     625
13    183
12      2
Name: count, dtype: int64

In [150]:
col = "telefono_1"
mask = df[col].str.len().ne(13)

display(df.loc[mask, ["nif", col]])

,nif,telefono_1
200,G38891446,(34)62573831
396,G76365618,(34)6260101199
587,G38764999,(34)92315082
620,G09793431,(34)+34650393946
667,G35385525,(34)6639474014
747,G56618713,(34)6
783,G19965037,(34)6296477


In [151]:
col = "telefono_2"
mask = df[col].ne("_U") & df[col].str.len().ne(13)

display(df.loc[mask, ["nif", col]])

,nif,telefono_2
185,G35671940,(34)63979131
233,G38829396,(34)63662970


In [152]:
pattern = r"\(34\)[0-9]{9}"

for col in ["telefono_1", "telefono_2"]:
    phones = df[col]

    has_value = phones.notna() & phones.ne("_U")
    matches = phones.str.fullmatch(pattern, na=False)

    print(col)
    display(df.loc[has_value & ~matches, ["nif", col]])

telefono_1


,nif,telefono_1
200,G38891446,(34)62573831
396,G76365618,(34)6260101199
587,G38764999,(34)92315082
620,G09793431,(34)+34650393946
667,G35385525,(34)6639474014
747,G56618713,(34)6
783,G19965037,(34)6296477


telefono_2


,nif,telefono_2
185,G35671940,(34)63979131
233,G38829396,(34)63662970


In [153]:
same_phone = (
    df["telefono_1"].notna()
    & df["telefono_2"].notna()
    & df["telefono_1"].ne("_U")
    & df["telefono_2"].ne("_U")
    & df["telefono_1"].eq(df["telefono_2"])
)

print("The same phone in both fields:", same_phone.sum())

The same phone in both fields: 20


In [154]:
phones_per_nif = (
    df.loc[
        df["telefono_1"].notna() & df["telefono_1"].ne("_U")
    ]
    .groupby("nif")["telefono_1"]
    .nunique()
)

display(phones_per_nif.value_counts())

telefono_1
1    745
Name: count, dtype: int64

In [155]:
phones_per_nif = (
    df.loc[
        df["telefono_2"].notna() & df["telefono_2"].ne("_U")
    ]
    .groupby("nif")["telefono_2"]
    .nunique()
)

display(phones_per_nif.value_counts())

telefono_2
1    172
Name: count, dtype: int64

### Telephone fields — conclusions

- `telefono_1` and `telefono_2` are string fields and should remain strings.
- Both contain 810 non-null values, with no blank strings or leading/trailing
  whitespace.

#### `telefono_1`

- Contains 728 distinct values and no `_U` placeholders.
- 803 values match the pattern `(34)` followed by exactly 9 digits.
- 7 values do not match this pattern: observed issues include unexpected
  digit counts and a repeated country-code representation, `(34)+34…`.
- Each of the 745 distinct NIFs has exactly one distinct `telefono_1`
  value across its records.

#### `telefono_2`

- Contains 171 distinct values, including `_U`.
- `_U` occurs in 625 records; 185 records contain another value.
- Of those 185 values, 183 match the same telephone pattern and 2 do not.
- Among the 172 NIFs with a second telephone other than `_U`, each has
  exactly one distinct second telephone across its records.

#### Relationships and integration considerations

- In 20 records, `telefono_1` and `telefono_2` contain the same value;
  the second field therefore does not always provide an additional number.
- Pattern matching confirms formatting only, not telephone validity
  or reachability.
- Telephone numbers have not been established as unique organisation
  identifiers.
- Preserve source values and flag format anomalies for later review.
- Rules for representing both fields in `Entidad.contacto` will be
  defined during integration.
- These findings apply to the analysed dataset.

### Campos de teléfono — conclusiones

- `telefono_1` y `telefono_2` son campos de texto y deben conservarse como tales.
- Ambos contienen 810 valores no nulos, sin cadenas vacías ni espacios
  al principio o al final.

#### `telefono_1`

- Contiene 728 valores distintos y ningún marcador `_U`.
- 803 valores cumplen el patrón `(34)` seguido de exactamente 9 dígitos.
- 7 valores no cumplen el patrón: se observaron cantidades inesperadas
  de dígitos y una representación repetida del prefijo internacional,
  `(34)+34…`.
- Cada uno de los 745 NIF distintos tiene un único valor distinto
  de `telefono_1` en sus registros.

#### `telefono_2`

- Contiene 171 valores distintos, incluido `_U`.
- `_U` aparece en 625 registros; 185 registros contienen otro valor.
- De esos 185 valores, 183 cumplen el mismo patrón telefónico y 2 no.
- Entre los 172 NIF con un segundo teléfono diferente de `_U`, cada uno
  tiene un único segundo teléfono distinto en sus registros.

#### Relaciones y consideraciones para la integración

- En 20 registros, `telefono_1` y `telefono_2` contienen el mismo valor;
  el segundo campo no siempre aporta un número adicional.
- El cumplimiento del patrón confirma únicamente el formato, no la validez
  ni la disponibilidad del teléfono.
- No se ha establecido que los teléfonos identifiquen organizaciones
  de forma única.
- Deben conservarse los valores originales y señalarse las anomalías
  de formato para su revisión posterior.
- Las reglas para representar ambos campos en `Entidad.contacto`
  se definirán durante la integración.
- Estos resultados se refieren al conjunto de datos analizado.

In [156]:
col = "pagina_web"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
print("Blank values:", df[col].str.strip().eq("").sum())
print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))

dtype: str
missing: 0
non-null: 810
unique: 173
Placeholder _U: 619
Blank values: 0
Leading/trailing whitespace: 0


0               wwwadepsi.org
1                          _U
2         www.sjdlaspalmas.es
3    www.asociacioncivitas.es
4                          _U
5                          _U
6                          _U
7                          _U
8                          _U
9                          _U
Name: pagina_web, dtype: str

pagina_web
_U                                                 619
www.cardiolatidos.org                                3
www.maroparque.es                                    3
www.fundacioncof.org                                 3
facebook: asociación psicosocial tuvidacondolor      2
www.lares.org.es                                     2
www.acciónaborigen.com                               2
www.fcbelenmaria.com                                 2
www.boscoglobal.org                                  2
www.reconoce.org                                     2
ecoarmonia.org                                       2
www.innovaprojectsocial.es                           2
www.starup.es                                        2
https://www.irissocial.es/                           2
www.fundec.org                                       2
www.accem.es                                         2
www.culturasunidas.org                               2
wwwadepsi.org                                        1

In [157]:
websites = df.loc[df["pagina_web"].ne("_U"), "pagina_web"]

with pd.option_context("display.max_rows", None):
    display(websites.drop_duplicates().sort_values())

615                                    adeintenerife.org
576                                             afaes.es
726                                      aflordemayo.com
490                                           apaelp.org
451                             asociacionacompañame.com
781                          asociaciondigitalcanaria.es
686                                 asociacionhekate.org
298                              asociacionmihijoyyo.org
760                                  asociacionsilva.com
656             asociacionsocialsolidaridadvenezuela.org
436                               barriosorquestados.org
358                              blog apostolado del mar
446                          cadenadefavoreslagomera.com
725                                       cbcanarias.net
623                                  cerroscolorazul.com
534                                          diaconia.es
714                                       ecoarmonia.org
225                            

In [158]:
display(websites[websites.str.contains(r"\s", na=False)])

72                              www. raizdelpueblo.org
99                                 www. afalapalma.net
358                            blog apostolado del mar
466    facebook: asociación psicosocial tuvidacondolor
467    facebook: asociación psicosocial tuvidacondolor
563                       en migración a nuevo dominio
Name: pagina_web, dtype: str

In [159]:
websites_per_nif = (
    df.loc[df["pagina_web"].ne("_U")]
    .groupby("nif")["pagina_web"]
    .nunique()
)

display(websites_per_nif.value_counts())

pagina_web
1    172
Name: count, dtype: int64

### `pagina_web` — conclusions

- String field with 810 non-null values and 173 distinct values, including `_U`.
- `_U` occurs in 619 records. The remaining 191 records contain 172 distinct
  values other than `_U`.
- No blank strings or leading/trailing whitespace were detected.
- Six records contain internal whitespace: two address-like values and
  four records containing textual descriptions.
- The field mixes addresses with and without HTTP/HTTPS schemes, social media
  links, textual descriptions, and an email address.
- Observed issues include spaces after `www.`, the misspelled scheme `hhtps://`,
  and descriptions such as “en migración a nuevo dominio”.
- Some domain-like values appear suspicious, but corrections have not
  been verified. Accented characters alone do not establish an error.
- Among the 172 NIFs with a value other than `_U`, each corresponds
  to exactly one distinct `pagina_web` value.
- Website syntax and availability have not been comprehensively validated.
- Mapping to `Entidad.url` requires distinguishing addresses from other
  content and defining cleaning rules. Preserve original values for reference.
- These findings apply to the analysed dataset.

### `pagina_web` — conclusiones

- Campo de texto con 810 valores no nulos y 173 valores distintos, incluido `_U`.
- `_U` aparece en 619 registros. Los 191 registros restantes contienen
  172 valores distintos diferentes de `_U`.
- No se detectaron cadenas vacías ni espacios al principio o al final.
- Seis registros contienen espacios internos: dos valores con apariencia
  de dirección web y cuatro registros con descripciones textuales.
- El campo mezcla direcciones con y sin protocolo HTTP/HTTPS, enlaces
  a redes sociales, descripciones textuales y una dirección de correo electrónico.
- Se observaron espacios después de `www.`, el protocolo mal escrito `hhtps://`
  y descripciones como “en migración a nuevo dominio”.
- Algunos valores con apariencia de dominio resultan sospechosos, pero
  las correcciones no se han verificado. Los caracteres acentuados
  no demuestran por sí solos un error.
- Entre los 172 NIF con un valor diferente de `_U`, cada uno corresponde
  a un único valor distinto de `pagina_web`.
- No se han validado exhaustivamente la sintaxis ni la disponibilidad
  de los sitios.
- La asociación con `Entidad.url` requiere distinguir las direcciones
  de otros contenidos y definir reglas de limpieza. Deben conservarse
  los valores originales como referencia.
- Estos resultados se refieren al conjunto de datos analizado.

In [161]:
col = "area_id"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
##print("Blank values:", df[col].str.strip().eq("").sum())
##print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))

dtype: int64
missing: 0
non-null: 810
unique: 9
Placeholder _U: 0


0    3
1    7
2    3
3    3
4    2
5    2
6    2
7    2
8    2
9    1
Name: area_id, dtype: int64

area_id
8    414
9    104
2     98
3     73
1     57
5     38
6     23
4      2
7      1
Name: count, dtype: int64

In [162]:
col = "area_nombre"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
print("Blank values:", df[col].str.strip().eq("").sum())
print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))

dtype: str
missing: 0
non-null: 810
unique: 9
Placeholder _U: 0
Blank values: 0
Leading/trailing whitespace: 0


0             Discapacidad
1    Centros Ocupacionales
2             Discapacidad
3             Discapacidad
4             Tercera Edad
5             Tercera Edad
6             Tercera Edad
7             Tercera Edad
8             Tercera Edad
9       Menores Y Juventud
Name: area_nombre, dtype: str

area_nombre
Varios                   414
Voluntariado             104
Tercera Edad              98
Discapacidad              73
Menores Y Juventud        57
Mujer                     38
Exclusión Social          23
Drogodependencia           2
Centros Ocupacionales      1
Name: count, dtype: int64

In [163]:
df.groupby("area_id")[
    "area_nombre"
].nunique().value_counts()

area_nombre
1    9
Name: count, dtype: int64

In [164]:
df.groupby("area_nombre")[
    "area_id"
].nunique().value_counts()

area_id
1    9
Name: count, dtype: int64

In [165]:
display(
    df[["area_id", "area_nombre"]]
    .drop_duplicates()
    .sort_values("area_id")
)

,area_id,area_nombre
9,1,Menores Y Juventud
4,2,Tercera Edad
0,3,Discapacidad
142,4,Drogodependencia
128,5,Mujer
10,6,Exclusión Social
1,7,Centros Ocupacionales
11,8,Varios
162,9,Voluntariado


In [166]:
df.groupby("nif")["area_id"].nunique().value_counts()

area_id
1    729
2     14
3      1
5      1
Name: count, dtype: int64

In [170]:
areas_per_nif = df.groupby("nif")["area_nombre"].unique()
with pd.option_context("display.max_colwidth", None):
    display(areas_per_nif[areas_per_nif.str.len() > 1])

nif
G13426671                                                    [Varios, Mujer]
G16824088                                          [Mujer, Exclusión Social]
G16979866                                   [Tercera Edad, Exclusión Social]
G19836733                                             [Discapacidad, Varios]
G19853902                                         [Exclusión Social, Varios]
G35141365                     [Exclusión Social, Voluntariado, Discapacidad]
G44942555                                         [Drogodependencia, Varios]
G67792655                                                    [Varios, Mujer]
G67946814                                 [Menores Y Juventud, Discapacidad]
G72405277    [Menores Y Juventud, Mujer, Discapacidad, Varios, Tercera Edad]
G76191220                                       [Varios, Menores Y Juventud]
G76285683                                         [Exclusión Social, Varios]
G76365410                                             [Varios, Tercera E

### `area_id` and `area_nombre` — conclusions

- `area_id` is stored as `int64` and represents a category code.
  `area_nombre` is a string field describing the activity category.
- There are 9 distinct codes and 9 distinct names.
- No null values were detected in either field. No `_U` placeholders,
  blank strings, or leading/trailing whitespace were detected in `area_nombre`.
- In this dataset, `area_id` and `area_nombre` have a one-to-one relationship:
  each code corresponds to exactly one name and vice versa.

#### Relationship with organisations

- Of the 745 distinct NIFs:
  - 729 have one activity category;
  - 14 have two;
  - 1 has three;
  - 1 has five.
- Therefore, 16 organisations have multiple activity categories.
- The observed relationship between `nif` and activity category is many-to-many:
  an organisation may have multiple categories, and a category may apply
  to multiple organisations.

#### Implications for later integration

- `area_nombre` is a candidate source for `Entidad.ambito`, but source
  categories require explicit mapping to the global `Ambito` enumeration.
- Categories such as `Varios` do not provide an unambiguous global category.
- If `Entidad.ambito` accepts only one value, an integration rule or schema
  adjustment is needed to preserve multiple activity categories.
- Preserve all observed categories until mapping rules are defined.
- These findings apply to the analysed dataset.

### `area_id` y `area_nombre` — conclusiones

- `area_id` está almacenado como `int64` y representa un código de categoría.
  `area_nombre` es un campo de texto que describe el área de actividad.
- Hay 9 códigos distintos y 9 denominaciones distintas.
- No se detectaron valores nulos en ninguno de los campos. En `area_nombre`
  tampoco se detectaron marcadores `_U`, cadenas vacías ni espacios
  al principio o al final.
- En este conjunto de datos, `area_id` y `area_nombre` tienen una relación
  de uno a uno: cada código corresponde a un único nombre y viceversa.

#### Relación con las organizaciones

- De los 745 NIF distintos:
  - 729 tienen un área de actividad;
  - 14 tienen dos;
  - 1 tiene tres;
  - 1 tiene cinco.
- Por tanto, 16 organizaciones tienen varias áreas de actividad.
- La relación observada entre `nif` y área de actividad es de varios a varios:
  una organización puede tener varias áreas y una misma área puede corresponder
  a varias organizaciones.

#### Implicaciones para la integración posterior

- `area_nombre` es un campo candidato para `Entidad.ambito`, pero las categorías
  de origen requieren una correspondencia explícita con la enumeración
  global `Ambito`.
- Categorías como `Varios` no permiten determinar una categoría global
  de forma inequívoca.
- Si `Entidad.ambito` admite un único valor, será necesaria una regla
  de integración o una adaptación del esquema para conservar varias áreas.
- Deben conservarse todas las categorías observadas hasta definir
  las reglas de correspondencia.
- Estos resultados se refieren al conjunto de datos analizado.

In [172]:
col = "subarea_id"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
print("Blank values:", df[col].str.strip().eq("").sum())
print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))

dtype: str
missing: 0
non-null: 810
unique: 17
Placeholder _U: 394
Blank values: 0
Leading/trailing whitespace: 0


0    _U
1    _U
2    _U
3    _U
4    _U
5    _U
6    _U
7    _U
8    _U
9    _U
Name: subarea_id, dtype: str

subarea_id
_U    394
B      90
E      86
I      50
H      41
F      22
5      21
4      20
C      15
7      13
6      13
G      12
2      11
D      10
3       8
A       2
1       2
Name: count, dtype: int64

In [173]:
col = "subarea_nombre"

print("dtype:", df[col].dtype)
print("missing:", df[col].isna().sum())
print("non-null:", df[col].notna().sum())
print("unique:", df[col].nunique())

print("Placeholder _U:", df[col].eq("_U").sum())
print("Blank values:", df[col].str.strip().eq("").sum())
print("Leading/trailing whitespace:", df[col].ne(df[col].str.strip()).sum())

display(df[col].dropna().head(10))
display(df[col].value_counts(dropna=False).head(20))

dtype: str
missing: 0
non-null: 810
unique: 17
Placeholder _U: 394
Blank values: 0
Leading/trailing whitespace: 0


0    _U
1    _U
2    _U
3    _U
4    _U
5    _U
6    _U
7    _U
8    _U
9    _U
Name: subarea_nombre, dtype: str

subarea_nombre
_U                                                                                                                                                                                       394
Comedores Sociales Casas Hogar Comunidades Religiosas Talleres Sociales Desarrollo Social E Inserción Social Y Fomento Y Empleo                                                           90
Asociaciones Culturales Deportivas SociO-Educativas SociO-Culturales                                                                                                                      86
Otros                                                                                                                                                                                     50
Asociaciones De Enfermos Ongs Relacionadas Con El Área SociO-Sanitaria Organizaciones Relacionadas Con Autoayuda Terapias Y Salud                                                         41
Asocaciones Relacionadas Con Ciencia Tec

In [176]:
df.groupby("subarea_nombre")[
    "subarea_id"
].nunique().value_counts()

subarea_id
1    17
Name: count, dtype: int64

In [177]:
df.groupby("subarea_id")[
    "subarea_nombre"
].nunique().value_counts()

subarea_nombre
1    17
Name: count, dtype: int64

In [175]:
display(
    df[["area_id","subarea_id", "subarea_nombre"]]
    .drop_duplicates()
    .sort_values("subarea_id")
)

,area_id,subarea_id,subarea_nombre
439,9,1,Cooperación Y Solidaridad Internacional
391,9,2,Servicios Sociales
566,9,3,Asuntos Sanitarios
475,9,4,Educación Ciencia Cultura Deportes Y Patrimoni...
437,9,5,Protección Del Medio Ambiente Y Defensa Del Me...
470,9,6,Protección Civil
461,9,7,Cualquier Otro Campo De Actuación En El Que Se...
226,8,A,Asociaciones De Vecinos Y Juveniles
34,8,B,Comedores Sociales Casas Hogar Comunidades Rel...
67,8,C,Asociación De Igualdad Lgtbi


In [178]:
subareas_per_group = (
    df.loc[df["subarea_id"].ne("_U")]
    .groupby(["nif", "area_id"])["subarea_id"]
    .nunique()
)

display(subareas_per_group.value_counts())

subarea_id
1    341
2     29
3      3
6      1
Name: count, dtype: int64

### `subarea_id` and `subarea_nombre` — conclusions

- Both fields contain 810 non-null string values.
- No blank strings or leading/trailing whitespace were detected.
- Each field has 17 distinct values: 16 subarea codes/names and `_U`.
- The marker pair `_U → _U` occurs in 394 records. Its exact meaning
  (unknown, unspecified, or not applicable) remains unconfirmed.
- In this dataset, `subarea_id` and `subarea_nombre` have a one-to-one
  relationship, including the marker pair.

#### Relationship with activity areas

- Subarea codes `1`–`7` occur only within `area_id = 9` (`Voluntariado`).
- Subarea codes `A`–`I` occur only within `area_id = 8` (`Varios`).
- Each known subarea code belongs to exactly one activity area in this dataset.
- `_U` occurs in all nine activity areas and must not be treated as a subarea.

#### Relationship with organisations

After excluding `_U`, grouping by `nif` and `area_id` shows:

- 341 organisation–area groups have one distinct subarea;
- 29 have two;
- 3 have three;
- 1 has six.

Therefore, 33 organisation–area groups have multiple subareas.
These counts describe groups, not necessarily distinct organisations.
Rows sharing a NIF and activity area may contain different subarea information
and should not be removed solely on that basis.

#### Implications for later integration

- Subareas provide additional detail that may help map broad categories
  such as `Varios` to the global `Ambito` enumeration.
- Preserve all known subareas until mapping rules are defined.
- These findings apply to the analysed dataset.

### `subarea_id` y `subarea_nombre` — conclusiones

- Ambos campos contienen 810 valores de texto no nulos.
- No se detectaron cadenas vacías ni espacios al principio o al final.
- Cada campo tiene 17 valores distintos: 16 códigos/nombres de subáreas y `_U`.
- La pareja de marcadores `_U → _U` aparece en 394 registros. Su significado
  exacto (desconocido, no especificado o no aplicable) no está confirmado.
- En este conjunto de datos, `subarea_id` y `subarea_nombre` tienen una relación
  de uno a uno, incluida la pareja de marcadores.

#### Relación con las áreas de actividad

- Los códigos de subárea `1`–`7` aparecen únicamente en `area_id = 9`
  (`Voluntariado`).
- Los códigos `A`–`I` aparecen únicamente en `area_id = 8` (`Varios`).
- Cada código de subárea conocido pertenece a una única área de actividad
  en este conjunto de datos.
- `_U` aparece en las nueve áreas y no debe interpretarse como una subárea.

#### Relación con las organizaciones

Tras excluir `_U`, la agrupación por `nif` y `area_id` muestra:

- 341 grupos organización–área tienen una subárea distinta;
- 29 tienen dos;
- 3 tienen tres;
- 1 tiene seis.

Por tanto, 33 grupos organización–área tienen varias subáreas.
Estas cifras describen grupos, no necesariamente organizaciones distintas.
Las filas que comparten NIF y área pueden aportar diferentes subáreas
y no deben eliminarse únicamente por compartir esos campos.

#### Implicaciones para la integración posterior

- Las subáreas aportan detalles que pueden ayudar a asociar categorías
  generales como `Varios` con la enumeración global `Ambito`.
- Deben conservarse todas las subáreas conocidas hasta definir
  las reglas de correspondencia.
- Estos resultados se refieren al conjunto de datos analizado.